library/src/main/java/com/devbrackets/android/exomedia/core/ListenerMux.kt

```diff
@@ -0,0 +1,460 @@
+import com.google.android.exoplayer2.ExoPlaybackException
+import com.google.android.exoplayer2.Format
+import com.google.android.exoplayer2.PlaybackParameters
+import com.google.android.exoplayer2.Player
+import com.google.android.exoplayer2.analytics.AnalyticsListener
+import com.google.android.exoplayer2.audio.AudioAttributes
+import com.google.android.exoplayer2.decoder.DecoderCounters
+import com.google.android.exoplayer2.metadata.Metadata
+import com.google.android.exoplayer2.source.MediaSourceEventListener
+import com.google.android.exoplayer2.source.TrackGroupArray
+import com.google.android.exoplayer2.trackselection.TrackSelectionArray
+import java.io.IOException
+import java.lang.ref.WeakReference
+
+/**
+ * An internal Listener that implements the listeners for the [ExoMediaPlayer],
+ * Android VideoView, and the Android MediaPlayer to output to the correct
+ * error listeners.
+ */
+class ListenerMux(private val muxNotifier: Notifier) : ExoPlayerListener, MediaPlayer.OnPreparedListener, MediaPlayer.OnCompletionListener, MediaPlayer.OnErrorListener, MediaPlayer.OnBufferingUpdateListener, MediaPlayer.OnSeekCompleteListener, OnBufferUpdateListener, MetadataListener, AnalyticsListener {
+
+    private val delayedHandler = Handler()
+
+    private var preparedListener: OnPreparedListener? = null
+    private var completionListener: OnCompletionListener? = null
+    private var bufferUpdateListener: OnBufferUpdateListener? = null
+    private var seekCompletionListener: OnSeekCompletionListener? = null
+    private var errorListener: OnErrorListener? = null
+    private var metadataListener: MetadataListener? = null
+    private var analyticsListener: AnalyticsListener? = null
+
+    private var clearableSurfaceRef = WeakReference<ClearableSurface>(null)
+
+    /**
+     * Retrieves if the player was prepared
+     *
+     * @return True if the player was prepared
+     */
+    var isPrepared = false
+        private set
+    private var notifiedCompleted = false
+    private var clearRequested = false
+
+    override fun onBufferingUpdate(mp: MediaPlayer, percent: Int) {
+        onBufferingUpdate(percent)
+    }
+
+    override fun onCompletion(mp: MediaPlayer) {
+        completionListener?.onCompletion()
+    }
+
+    override fun onError(mp: MediaPlayer, what: Int, extra: Int): Boolean {
+        return notifyErrorListener(NativeMediaPlaybackException(what, extra))
+    }
+
+    override fun onSeekComplete(mp: MediaPlayer) {
+        seekCompletionListener?.onSeekComplete()
+    }
+
+    override fun onPrepared(mp: MediaPlayer) {
+        notifyPreparedListener()
+    }
+
+    override fun onError(exoMediaPlayer: ExoMediaPlayer, e: Exception?) {
+        muxNotifier.onMediaPlaybackEnded()
+        muxNotifier.onExoPlayerError(exoMediaPlayer, e)
+        notifyErrorListener(e)
+    }
+
+    override fun onStateChanged(playWhenReady: Boolean, playbackState: Int) {
+        if (playbackState == Player.STATE_ENDED) {
```

Comment:
Using `when` clauses might be clearer? Especially since you are checking states with equality, some of these branches seem awkwardly placed. (why no `else` in the later statements?)